### PR TITLE
[CURA-9036] Fix using wrong junctions when calculating areas.

### PR DESCRIPTION
**Pull request by rburema**
The code as it was assumed that each junction specified the width for the _previous_ line it was connected to, but it actually specifies the width for the _next_ line it's connected to.

This caused an issue where it could make the symmetrical beadings so far out of kilter that it's easily visible in the preview.

Technically this still can happen, but only because the Maximum Extrusion Area Deviation is set to a very large value in the front-end, so the simplification doesn't really care about the weights. We should probably put it back to 2000 square micron, as it was.

Also remove that we re-set the width of the next-vertex when simplifying out the current, since this can have repercussions for line-segments that have very little to do with the removal of the current vertex in the loop. (Namely the next-next.)

See also frontend PR https://github.com/Ultimaker/Cura/pull/11709 since we need the maximum area deviation back at a somewhat more reasonable figure to make this work at all.

### CODE REVIEW

**Comment by Ghostkeeper** [comment]
Some of the ExtrusionLine tests are currently failing. We'll have to take a look at which is correct; the code or the tests.

**Review by Ghostkeeper — CHANGES_REQUESTED** [review]

File: src/utils/ExtrusionLine.cpp [hunk 1]
```diff
@@ -220,18 +220,18 @@ coord_t ExtrusionLine::calculateExtrusionAreaDeviationError(ExtrusionJunction A,
      * */
     const coord_t ab_length = vSize(B - A);
     const coord_t bc_length = vSize(C - B);
-    const coord_t width_diff = llabs(B.w - C.w);
+    const coord_t width_diff = llabs(B.w - A.w);
```
[the remark this answers is not in the document]
Comment (rburema) [reply src/utils/ExtrusionLine.cpp:223]: Ah I missed that. When I write things myself I already use that.

File: src/utils/ExtrusionLine.cpp [hunk 2]
```diff
@@ -220,18 +220,18 @@ coord_t ExtrusionLine::calculateExtrusionAreaDeviationError(ExtrusionJunction A,
...
-        return llabs(B.w - weighted_average_width) * ab_length + llabs(C.w - weighted_average_width) * bc_length;
+        weighted_average_width = (ab_length * A.w + bc_length * B.w) / vSize(C - A);
+        return llabs(A.w - weighted_average_width) * ab_length + llabs(B.w - weighted_average_width) * bc_length;
     }
     else
     {
         // If the width difference is very small, then select the width of the segment that is longer
-        weighted_average_width = ab_length > bc_length ? B.w : C.w;
+        weighted_average_width = ab_length > bc_length ? A.w : B.w;
         return ab_length > bc_length ? width_diff * bc_length : width_diff * ab_length;
```
[the remark this answers is not in the document]
Comment (rburema) [reply src/utils/ExtrusionLine.cpp:235]: I don't think so; the result of the function is supposed to be the difference in area caused by the proposed merge. Which is the difference times the length of the one you _don't_ keep.

File: src/utils/ExtrusionLine.cpp [hunk 3]
```diff
@@ -220,18 +220,18 @@ coord_t ExtrusionLine::calculateExtrusionAreaDeviationError(ExtrusionJunction A,
...
     const coord_t bc_length = vSize(C - B);
-    const coord_t width_diff = llabs(B.w - C.w);
+    const coord_t width_diff = llabs(B.w - A.w);
     if (width_diff > 1)
     {
         // Adjust the width only if there is a difference, or else the rounding errors may produce the wrong
         // weighted average value.
-        weighted_average_width = (ab_length * B.w + bc_length * C.w) / vSize(C - A);
-        return llabs(B.w - weighted_average_width) * ab_length + llabs(C.w - weighted_average_width) * bc_length;
+        weighted_average_width = (ab_length * A.w + bc_length * B.w) / vSize(C - A);
```
[the remark this answers is not in the document]
Comment (rburema) [reply src/utils/ExtrusionLine.cpp]: That works better: We can even up the Maximal Area Deviation from 2000 to 5000 micron, and the result still looks correct on visual inspection :-)

**Review by Ghostkeeper — APPROVED** [review]
Looks much more logical this way. Nice!

**Pull request merged**